Benchmarking

benjeffery opened the pull request:
WIP #2444 

Needs some polish, but probably a good point to get early feedback on the direction. The results are appended to a JSON file that will be committed to the repo, and a very basic HTML report generated:

![Screenshot from 2022-08-01 12-05-22](https://user-images.githubusercontent.com/8552/182134928-a01c15cc-87ed-415a-a402-abc3a498c048.png)

My thinking is that at under a minute we can add this to CI.

jeromekelleher commented:
Looks great!

Some unexpected differences here in the e.g. ``tree.parent_array``, which make me wonder if anything under 50ns should just be "green". Do you think these are significant?

Re running in CI, I guess we just run it to make sure it hasn't broken but don't pay any attention to the numbers? 

Probably worth spinning into its own Workflow, so we're not running it n-times in the Tests?

benjeffery commented:
> Looks great!
> 
> Some unexpected differences here in the e.g. `tree.parent_array`, which make me wonder if anything under 50ns should just be "green". Do you think these are significant?

Yes, the colouring was a quick hack. I think for those fast ones that is usual variance.
> 
> Re running in CI, I guess we just run it to make sure it hasn't broken but don't pay any attention to the numbers?
Exactly, mostly run it to check we didn't break it, occasionally look at the numbers if useful.

> Probably worth spinning into its own Workflow, so we're not running it n-times in the Tests?

Yeah, very simple action that will finish before all the other tests, so not making our CI longer in total.

benjeffery commented:
Ok, this needs some docs, but is ready for a code review.
Here is the full table - there are a couple of regressions that might be worth looking into.
![screencapture-file-home-benj-projects-tskit-python-benchmark-bench-results-html-2022-08-02-12_20_56](https://user-images.githubusercontent.com/8552/182363035-2be766ce-0a37-40c7-bf87-8e6bad29ebda.png)

jeromekelleher reviewed (approved):
Looks great!

The only major thing that's missing from the benchmarks is row-by-row access with/out metadata decoding, but we can add that in later.

petrelharp commented:
Wow!!!! This is great!

benjeffery commented:
I've added a test for decoding metadata. I've also added a script that runs the benchmarks across all released versions.
I've also added the script to CI (takes ~5m), the results are available as an artifact - they won't be directly comparable as it is a different CPU, but the relative changes should still be useful. I think this is ready to merge.

benjeffery commented:
BTW, this shows how great the backwards compatibility has been. Same 0.5.2 file for all these benchmarks!

jeromekelleher commented on python/benchmark/run-for-all-releases.py:
```diff
@@ -0,0 +1,19 @@
+if __name__ == "__main__":
+    versions = [v for v in versions("tskit") if "a" not in v and "b" not in v]
+    for v in tqdm.tqdm(versions):
+        os.system(f"pip install tskit=={v}")
```
``subprocess.run(check=True, shell=True)`` is a better option here

jeromekelleher replied:
I think it's probably worth automating in a venv here, so that the devs environment doesn't get messed up by running this.

jeromekelleher replied:
``subprocess.run(check=True, shell=True)`` is a better option here

jeromekelleher commented on python/benchmark/run.py:
```diff
@@ -0,0 +1,210 @@
+    return ret
+
+
+def make_file():
```
Slightly nicer:
```
benchfile = tskit_dir / "benchmark" / "bench.trees"
if not benchfile.exists():
     ...
     ts.dump(benchfile)
```
similar patterns below

jeromekelleher reviewed (approved):
Awesome!

Few minor suggestions, merge away whenever you're happy.
